src/engine.rs

```diff
@@ -679,22 +675,24 @@ impl Reedline {
             let mut paste_enter_state = false;
 
             #[cfg(feature = "external_printer")]
-            if let Some(ref external_printer) = self.external_printer {
-                // get messages from printer as crlf separated "lines"
-                let messages = Self::external_messages(external_printer)?;
-                if !messages.is_empty() {
-                    // print the message(s)
-                    self.painter.print_external_message(
-                        messages,
-                        self.editor.line_buffer(),
-                        prompt,
-                    )?;
+            if let Some(printer) = &self.external_printer {
+                let any_messages = self.painter.print_external_messages(
+                    printer.receiver(),
+                    self.editor.line_buffer(),
+                    prompt,
+                )?;
+
+                if any_messages {
                     self.repaint(prompt)?;
                 }
             }
 
             let mut latest_resize = None;
-            loop {
+
+            // There could be multiple events queued up!
+            // pasting text, resizes, blocking this thread (e.g. during debugging)
+            // We should be able to handle all of them as quickly as possible without causing unnecessary output steps.
+            while event::poll(Duration::from_millis(POLL_WAIT))? {
```

Comment:
There was some subtlety around when we `crossterm::event::poll` and `crossterm::event::read` to avoid idling in a hotter than necessary loop https://github.com/nushell/reedline/pull/651

So I am not sure if this `do ... while` to `while` transformation is correct.

Reply:
Unfortunately, the current loop structure will block on the first `read` and later external messages will not get printed until the user types which is confusing. I wonder if there's a better solution though...

Reply:
Damn you asynchronous concurrency....

Not blocking by polling certainly makes reading that channel possible.
Could we shove both kinds of events into one channel and block on that? (that would be the no async just threads everywhere path)

With how we dealt with `crossterm` and stdout so far I am certainly not screaming: `cargo add tokio`

Reply:
Using the channel + threads should work, but I think means there will have to be 3 threads in total:
- One to forward messages from `event::read`
- One to forward messages from the external printer
- The main thread, to pull messages off the channel

It is possible to remove the external printer thread if we wrap the external printer sender in a new type. This way, we can allow the external thread to send messages directly to the merged channel (`Reciever<ReedlineMessage>`), instead of having to move data from the external channel `Reciever<Vec<u8>>` to the merged channel. E.g.,:
```
enum ReedlineMessage {
    Event(crossterm::Event),
    ExternalMessage(Vec<u8>),
}

struct ExternalPrinterSender(SyncSender<ReedlineMessage>)

impl ExternalPrinterSender {
    pub fn send(&self, data: Vec<u8>) -> SendError<Vec<u8> {
        self.0.send(ReedlineMessage::ExternalMessage(data)).map_err(...);
    }
}
```
The potential problem with this is that `nu-system`, or wherever job controls ends up, would need to take `reedline` as a dependency.

Reply:
Crossterm mentions a restriction on the threads their blocking functions can be run from and that it is incompatible with their async oriented `EventStream`

https://docs.rs/crossterm/latest/crossterm/event/index.html

Maybe we need to bite the `Future` bullet at some point? (providing an executor would suck for library users, but we may want to unlock things like async prompt or completion updates.)

Reply:
> Crossterm mentions a restriction on the threads their blocking functions can be run from

This should be fine, one thread can both poll and read.

> Maybe we need to bite the Future bullet at some point?

In the future (ha), it might be necessary. So far, it looks like we can probably work around it?

Reply:
Two other possible solutions:
1. Use a higher poll wait time like it was suggested in #757. We could use the higher wait time for the first poll only and then switch back to the current wait time after the first read.
2. Have only two threads:
   1. the main thread: has the old `poll` + `read` loop and prints the prompt, etc.
   2. the external printer thread: takes messages off the channel receiver and prints these messages.
   
   Then, we synchronize the two threads using a `Mutex` or something similar to ensure that only one of them prints to the terminal at a time.

Reply:
Not sure how 1) would play out both in terms of external latency and syscall churn. (would need to see it)

2) sounds more involved. Question would be if the `crossterm` cursor or terminal size polling operations may collide with streaming output from another thread. But yeah with the appropriate locking this may be workable.